### DISCUSSION TITLE
Accessing previous state when updating state

### DISCUSSION THREAD

**Discussion by Ettur** [pmndrs/zustand#821] [Q&A] [answered]
Hi,

This is not really an issue but I did not know where can I ask this information so I put it here.
Can I, and if so how, access previous state when updating state ?
What I mean is that for example with basic React useState i can write

```
const [test, setTest] = useState ({first: true, second: true});
setTest (prev => {return {...prev, second: false }});
```
here I can access previous state and update only 'second' property;
with zustand can I do the same somehow ?
for example in zustand store i have

```
interface View {
    current: string,
    next: boolean
}

interface AppState {
    view: View;
    setView: (view: View) => void;
}

export const useStore = create<AppState>((set: SetState<AppState>, get: GetState<AppState>) => ({   
    view: {current: 'landing', next: false},
    setView: (view: View) => {set ({view: view})},
}));

```

and then I consume store and try to update in component

```
const setView = useStore(state => state.setView);
setView( prev => ...prev) // THIS HERE!! How can I access previous state ? for example to only update next property while using current from previous state.

```

**Comment by dai-shi** [comment, marked as answer]
This is what you want.
```js
export const useStore = create<AppState>((set: SetState<AppState>, get: GetState<AppState>) => ({   
  view: {current: 'landing', next: false},
  setView: (fn: (prev: View) => View) => {
    set((state => ({ view: fn(state.view) });
  },
}));
```

You could also do `useStore.getState()` and `useStore.setState(...)` if you want.

**Reply by xXAvoraXx** [reply]
Does the get function get the state after the state is updated or before the state is updated?

**Reply by dai-shi** [reply]
It returns undefined during initialization. You can try it. Is it what you are asking?

**Reply by xXAvoraXx** [reply]
For example, examine the code below.

```
...create((set, get) =>
...
removeNode: (id) => {
const oldNodes = get().nodes;
      set({
        nodes: oldNodes.filter((x) => x.id !== node.id),
      });
const nodes = get().nodes;
}
```

After the set function updates the state, if I call the state with the get function, will I get the updated state or the old state?

**Reply by dai-shi** [reply]
Yes, that should work fine. You get updated one.

**Reply by huntzinger92** [reply]
This was a great reference, thank you. In my case, I wanted the state setter to work like a React state setter, where consuming code can pass a new state value OR a state setting function that uses previous state. I think the above forces you to use the state setting function.

Here is a generic solution that allows you to treat zustand store state setters like React state setters, where you can pass the state value itself or a function that accesses previous state and returns a new state value:

```
import { isArray } from "lodash-es";
import { create } from "zustand";

// accept argument of type T or a callback using previous state of type T
type ReactStyleStateSetter<T> = T | ((prev: T) => T);

interface ExampleStoreState {
  arr: number[]; // or whatever your state type is
  setArr: (
    newArrOrSetterFn: ReactStyleStateSetter<number[]>
  ) => void;
}

export const useExampleStore = create<ExampleStoreState>()(
  (set) => ({
    arr: [],
    setArr: (newArrOrSetterFn) => {
      set(({ arr }) => {
        // your type check equivalent here
        if (isArray(newArrOrSetterFn)) {
          const newArr = newArrOrSetterFn;
          return { selectedImageIds: newArr };
        }
        const setterFn = newArrOrSetterFn;
        return {
          arr: setterFn(arr)
        };
      });
    }
  })
);
```

Then you can call `setArr` like `setArr([1,2,3])` or `setArr((prev) => [...prev, 4])`.
